js/packages/openinference-instrumentation-langchain/src/tracer.ts

```diff
@@ -0,0 +1,107 @@
+import { BaseTracer, Run } from "@langchain/core/tracers/base";
+import {
+  Tracer,
+  SpanKind,
+  Span,
+  context,
+  trace,
+  SpanStatusCode,
+} from "@opentelemetry/api";
+import { isTracingSuppressed } from "@opentelemetry/core";
+import { SemanticConventions } from "@arizeai/openinference-semantic-conventions";
+import {
+  safelyFlattenAttributes,
+  safelyFormatIO,
+  safelyFormatInputMessages,
+  safelyFormatOutputMessages,
+  safelyGetOpenInferenceSpanKindFromRunType,
+} from "./utils";
+
+type RunWithSpan = {
+  run: Run;
+  span: Span;
+};
+
+export class LangChainTracer extends BaseTracer {
+  private tracer: Tracer;
+  private runs: Record<string, RunWithSpan | undefined> = {};
+  constructor(tracer: Tracer) {
+    super();
+    this.tracer = tracer;
+  }
+  name: string = "OpenInferenceLangChainTracer";
+  protected persistRun(_run: Run): Promise<void> {
+    return Promise.resolve();
+  }
```

Comment:
this needed?

Reply:
yeah in order to extend BaseTracer, but i should call teh super in here so will do that

Reply:
actually BaseTracer is an abstract class, which cannot be instantiated but only extended / inherited from. the persist Run method is an abstract method which means it must be defined by whatever class inherits from the BaseTracer class since the BaseTracer class doesn't have it's own implementation of it.

https://www.tutorialsteacher.com/typescript/abstract-class so basically have to put something here but it's unused so just matched the type

Reply:
K, sg